// @liara/cli/lib/commands/app/list.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const core_1 = require("@oclif/core");
const base_1 = tslib_1.__importDefault(require("../../base"));
const shamsi = tslib_1.__importStar(require("shamsi-date-converter"));
class AppList extends base_1.default {
    async run() {
        const { flags } = await this.parse(AppList);
        await this.setGotConfig(flags);
        const { projects } = await this.got('v1/projects').json();
        if (projects.length === 0) {
            this.error("Please create an app via 'liara app:create' command, first.");
        }
        const appsData = projects.map((project) => {
            const shamshiDate = shamsi.gregorianToJalali(new Date(project.created_at));
            return {
                Name: project.project_id,
                Platform: project.type,
                Plan: project.planID,
                Status: project.status,
                Scale: project.scale,
                'Created At': `${shamshiDate[0]}-${shamshiDate[1]}-${shamshiDate[2]}`,
            };
        });
        core_1.CliUx.ux.table(appsData, {
            Name: {},
            Platform: {},
            Plan: {},
            Scale: {},
            Status: {},
            'Created At': {},
        }, flags);
    }
}
exports.default = AppList;
AppList.description = 'list available apps';
AppList.flags = Object.assign(Object.assign({}, base_1.default.flags), core_1.CliUx.ux.table.flags());
AppList.aliases = ['app:ls'];
