// @liara/cli/lib/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AVAILABLE_PLATFORMS = exports.MAX_SOURCE_SIZE = exports.FALLBACK_REGION = exports.REGIONS_API_URL = exports.GLOBAL_CONF_VERSION = exports.PREVIOUS_GLOBAL_CONF_PATH = exports.GLOBAL_CONF_PATH = exports.DEV_MODE = void 0;
const tslib_1 = require("tslib");
const os_1 = tslib_1.__importDefault(require("os"));
const path_1 = tslib_1.__importDefault(require("path"));
exports.DEV_MODE = process.argv.includes('--dev');
exports.GLOBAL_CONF_PATH = path_1.default.join(os_1.default.homedir(), '.liara-auth.json');
exports.PREVIOUS_GLOBAL_CONF_PATH = path_1.default.join(os_1.default.homedir(), '.liara.json');
exports.GLOBAL_CONF_VERSION = '1';
exports.REGIONS_API_URL = {
    iran: 'https://api.iran.liara.ir',
    germany: 'https://api.liara.ir',
};
exports.FALLBACK_REGION = 'iran';
exports.MAX_SOURCE_SIZE = 256 * 1024 * 1024; // 256 MiB
exports.AVAILABLE_PLATFORMS = [
    'node',
    'laravel',
    'php',
    'django',
    'flask',
    'netcore',
    'react',
    'angular',
    'vue',
    'static',
    'docker',
    'next',
];
