// @liara/cli/lib/commands/env/set.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const inquirer_1 = tslib_1.__importDefault(require("inquirer"));
const base_1 = tslib_1.__importDefault(require("../../base"));
const core_1 = require("@oclif/core");
const output_1 = require("../../utils/output");
class EnvSet extends base_1.default {
    async run() {
        const { flags, argv } = await this.parse(EnvSet);
        await this.setGotConfig(flags);
        const debug = output_1.createDebugLogger(flags.debug);
        if (argv.length === 0) {
            EnvSet.run(['-h']);
            this.exit(0);
        }
        const env = this.readKeyValue(argv);
        const app = flags.app || (await this.promptProject());
        const appliedEnvs = await this.fetchEnvs(app);
        const variables = [...appliedEnvs, ...env];
        try {
            if (flags.force || (await this.confirm())) {
                await this.got.post('v1/projects/update-envs', {
                    json: { project: app, variables },
                });
                this.log(`Configuration variable applied and restarting ${app}`);
            }
        }
        catch (error) {
            debug(error.message);
        }
    }
    async fetchEnvs(app) {
        const { project } = await this.got(`v1/projects/${app}`).json();
        const envs = project.envs.map((env) => {
            const key = env.key;
            const value = env.value;
            return { key, value };
        });
        return envs;
    }
    splitWithDelimiter(delimiter, string) {
        return (string.match(new RegExp(`(${delimiter}|[^${delimiter}]+)`, 'g')) || []);
    }
    removeFirstSyombol(splitedGroup) {
        let counter = 0;
        const removedOne = splitedGroup.map((item) => {
            let content = '';
            if (item === '=')
                counter++;
            counter === 1 && item === '=' ? false : (content += item);
            return content;
        });
        return removedOne;
    }
    readKeyValue(env) {
        const variable = env.map((env) => {
            const splitWithDelimiter = this.splitWithDelimiter('=', env);
            const removedOne = this.removeFirstSyombol(splitWithDelimiter).filter(Boolean);
            const [tKey, ...tValue] = removedOne;
            const [key, value] = [tKey, tValue.join('')];
            return { key, value };
        });
        return variable;
    }
    async confirm() {
        const { confirm } = (await inquirer_1.default.prompt({
            name: 'confirm',
            type: 'confirm',
            message: `Your app will be restarted due to these configuration changes. Confirm: `,
            default: false,
        }));
        return confirm;
    }
}
exports.default = EnvSet;
EnvSet.description = 'specifying environment variables to an app';
EnvSet.strict = false;
EnvSet.args = [
    {
        name: 'env',
        description: 'key=value pair',
    },
];
EnvSet.flags = Object.assign(Object.assign({}, base_1.default.flags), { app: core_1.Flags.string({ char: 'a', description: 'app id' }), force: core_1.Flags.boolean({ char: 'f', description: 'force update' }) });
