// @liara/cli/lib/commands/account/remove.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const fs_extra_1 = tslib_1.__importDefault(require("fs-extra"));
const chalk_1 = tslib_1.__importDefault(require("chalk"));
const base_1 = tslib_1.__importDefault(require("../../base"));
const inquirer_1 = require("inquirer");
const core_1 = require("@oclif/core");
const constants_1 = require("../../constants");
class AccountRemove extends base_1.default {
    async run() {
        const { flags } = await this.parse(AccountRemove);
        const liara_json = await this.readGlobalConfig();
        if (!liara_json ||
            !liara_json.accounts ||
            Object.keys(liara_json.accounts).length === 0) {
            this.error("Please add your accounts via 'liara account:add' command, first.");
        }
        const name = flags.account || (await this.promptName(liara_json.accounts));
        const selectedAccount = liara_json.accounts[name];
        !selectedAccount &&
            this.error(`Could not find any account associated with this name ${name}.`);
        const accounts = liara_json.accounts;
        delete accounts[name];
        const accountsLength = Object.keys(liara_json.accounts).length;
        fs_extra_1.default.writeFileSync(constants_1.GLOBAL_CONF_PATH, JSON.stringify({
            version: constants_1.GLOBAL_CONF_VERSION,
            accounts,
        }));
        this.log(chalk_1.default.red('Auth credentials removed.'));
        selectedAccount.current &&
            accountsLength > 0 &&
            this.log(chalk_1.default.cyan("Please select an acount via 'liara account:use' command."));
        accountsLength < 1 &&
            this.log(chalk_1.default.cyan("There are no more accounts to use. Please add an account via 'liara account:add' command."));
        const { accountName } = await this.getCurrentAccount();
        accountName && this.log(chalk_1.default.cyan(`Current account is: ${accountName}`));
    }
    async promptName(accounts) {
        const { name } = (await inquirer_1.prompt({
            name: 'name',
            type: 'list',
            message: 'Enter your account name:',
            choices: [...Object.keys(accounts)],
        }));
        return name;
    }
}
exports.default = AccountRemove;
AccountRemove.description = 'remove an account';
AccountRemove.flags = Object.assign(Object.assign({}, base_1.default.flags), { account: core_1.Flags.string({ char: 'a', description: 'account name' }) });
AccountRemove.aliases = ['account:rm'];
